sunpy/coordinates/spice.py

```diff
@@ -0,0 +1,217 @@
+from sunpy.time import parse_time
+from sunpy.time.time import _variables_for_parse_time_docstring
+from sunpy.util.decorators import add_common_docstring
+
+__all__ = ['get_body', 'initialize']
+
+
+# Note that this epoch is very slightly different from the typical definition of J2000.0 (in TT)
+_ET_REF_EPOCH = Time('J2000', scale='tdb')
+
+_CLASS_TYPES = {2: 'PCK', 3: 'CK', 4: 'TK', 5: 'dynamic', 6: 'switch'}
+
+
+# Registry of the generated frame classes
+spice_frame_classes = []
+
+
+# Defined for future use
+class _SpiceBaseCoordinateFrame(SunPyBaseCoordinateFrame):
+    pass
+
+
+def _convert_to_et(time):
+    return (time - _ET_REF_EPOCH).to_value('s')
+
+
+def _install_frame_by_id(frame_id):
+    frame_name = spiceypy.frmnam(frame_id)
+
+    # Replace plus/minus characters in the SPICE frame name with lowercase 'p'/'n'
+    frame_name = frame_name.translate(str.maketrans("+-", "pn"))
+
+    frame_center, class_num, _ = spiceypy.frinfo(frame_id)
+    frame_center_name = spiceypy.bodc2n(frame_center)
+    log.info(f"Installing {frame_name} {_CLASS_TYPES[class_num]} frame ({frame_id}) "
+             f"as 'spice_{frame_name}'")
+
+    spice_frame = type(f"spice_{frame_name}",
+                       (_SpiceBaseCoordinateFrame,),
+                       {})
+    # Force the capitalization pattern of lowercase "spice_" followed by uppercase SPICE frame name
+    spice_frame.name = spice_frame.__name__
+
+    @frame_transform_graph.transform(FunctionTransformWithFiniteDifference, ICRS, spice_frame)
+    def icrs_to_spice(from_icrs_coord, to_spice_frame):
+        et = _convert_to_et(to_spice_frame.obstime)
+        # .copy() is necessary due to a bug (https://github.com/astropy/astropy/issues/15503)
+        matrix = spiceypy.sxform('J2000', frame_name, et)[..., :3, :3].copy()
+        icrs_offset = spiceypy.spkpos(frame_center_name,
+                                      et, 'J2000', 'NONE', 'SSB')[0] << u.km
+        shifted_old_pos = from_icrs_coord.cartesian - CartesianRepresentation(icrs_offset.T)
+        new_pos = shifted_old_pos.transform(matrix)
+        return to_spice_frame.realize_frame(new_pos)
+
+    @frame_transform_graph.transform(FunctionTransformWithFiniteDifference, spice_frame, ICRS)
+    def spice_to_icrs(from_spice_coord, to_icrs_frame):
+        et = _convert_to_et(from_spice_coord.obstime)
+        # .copy() is necessary due to a bug (https://github.com/astropy/astropy/issues/15503)
+        matrix = spiceypy.sxform(frame_name, 'J2000', et)[..., :3, :3].copy()
+        shifted_new_pos = from_spice_coord.cartesian.transform(matrix)
+        icrs_offset = spiceypy.spkpos(frame_center_name,
+                                      et, 'J2000', 'NONE', 'SSB')[0] << u.km
+        new_pos = shifted_new_pos + CartesianRepresentation(icrs_offset.T)
+        return to_icrs_frame.realize_frame(new_pos)
+
+    frame_transform_graph._add_merged_transform(spice_frame, ICRS, spice_frame)
```

Comment:
Random aside, but we are making heavy use of this method in sunpy, should we open a PR to astropy to make it not private?

Reply:
oh maybe I am getting confused between your geo frames PR and the merged code lol. Either way, I think that if we are going to rely on it we should ask astropy to mark it as public, mainly so someone doesn't change it somehow thinking that's a safe thing to do?